torch_em/self_training/fix_match.py

```diff
@@ -0,0 +1,309 @@
+
+        # Check that we have at least one of supvervised / unsupervised val loader.
+        assert sum((
+            supervised_val_loader is not None,
+            unsupervised_val_loader is not None,
+        )) > 0
+        self.supervised_val_loader = supervised_val_loader
+        self.unsupervised_val_loader = unsupervised_val_loader
+
+        if self.unsupervised_val_loader is None:
+            val_loader = self.supervised_val_loader
+        else:
+            val_loader = self.unsupervised_train_loader
+
+        # Check that we have at least one of supvervised / unsupervised loss and metric.
+        assert sum((
+            supervised_loss_and_metric is not None,
+            unsupervised_loss_and_metric is not None,
+        )) > 0
+        self.supervised_loss_and_metric = supervised_loss_and_metric
+        self.unsupervised_loss_and_metric = unsupervised_loss_and_metric
+
+        super().__init__(
+            model=model, train_loader=train_loader, val_loader=val_loader,
+            loss=Dummy(), metric=Dummy(), logger=logger, **kwargs
+        )
+
+        self.unsupervised_loss = unsupervised_loss
+        self.supervised_loss = supervised_loss
+
+        self.pseudo_labeler = pseudo_labeler
+
+        if source_distribution is None:
+            self.source_distribution = None
+        else:
+            self.source_distribution = torch.FloatTensor(source_distribution).to(self.device)
+
+        self._kwargs = kwargs
+
+    def get_distribution_alignment(self, pseudo_labels, label_threshold=0.5):
+        if self.source_distribution is not None:
+            pseudo_labels_binary = torch.where(pseudo_labels >= label_threshold, 1, 0)
+            _, target_distribution = torch.unique(pseudo_labels_binary, return_counts=True)
+            target_distribution = target_distribution / target_distribution.sum()
+            distribution_ratio = self.source_distribution / target_distribution
+            pseudo_labels = torch.where(
+                pseudo_labels < label_threshold,
+                pseudo_labels * distribution_ratio[0],
+                pseudo_labels * distribution_ratio[1]
+            ).clip(0, 1)
+
+        return pseudo_labels
+
+    #
+    # training and validation functionality
+    #
+
+    def _train_epoch_unsupervised(self, progress, forward_context, backprop):
+        self.model.train()
+
+        n_iter = 0
+        t_per_iter = time.time()
+
+        # Sample from both the supervised and unsupervised loader.
+        for xu1, xu2 in self.unsupervised_train_loader:
+            xu1, xu2 = xu1.to(self.device), xu2.to(self.device)
+
+            teacher_input, model_input = xu1, xu2
+
+            self.optimizer.zero_grad()
+            # Perform unsupervised training
+            with forward_context():
+                # Compute the pseudo labels.
+                pseudo_labels, label_filter = self.pseudo_labeler(self.model, teacher_input)
```

Comment:
We need to be careful here that the pseudo labels are not taken into account when computing gradients. So this should be wrapped into a `with torch.no_grad()` and we should `detach` the `pseudo_labels`. (This might be missing from the mean teacher trainer as well and so should be added there too).

Reply:
Fixed it (in both FixMatch and MeanTeacher trainers). Thanks for mentioning!